agent/csi/volumes.go

```diff
@@ -0,0 +1,190 @@
+package csi
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/docker/swarmkit/agent/exec"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+)
+
+// volumes is a map that keeps all the currently available volumes to the agent
+// mapped by volume ID.
+type volumes struct {
+	mu     sync.RWMutex // To sync between Add(), Remove() and Reset() for "m" map
+	lock   sync.RWMutex // To sync between Get(), addVolumes() and removeVolumes()
+	m      map[string]*api.VolumeAssignment
+	plugin *NodePlugin
+}
+
+const MaxRetry int = 25
+
+// NewManager returns a place to store volumes.
+func NewManager() exec.VolumesManager {
+	return &volumes{
+		m: make(map[string]*api.VolumeAssignment),
+		plugin: &NodePlugin{
+			VolumeMap: make(map[string]*VolumePublishStatus),
+		},
+	}
+}
+
+// Get returns a volume published path for the provided volume ID.  If the volume doesn't exist, returns empty string.
+func (r *volumes) Get(volumeID string) (string, error) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	if volInfo, ok := r.plugin.VolumeMap[volumeID]; ok {
+		if volInfo.IsPublished {
+			return volInfo.TargetPath, nil
+		}
+		return volInfo.TargetPath, fmt.Errorf("volume %s is not published", volumeID)
+	}
+	return "", nil
+}
+
+// Add adds one or more volumes to the volume map.
+func (r *volumes) Add(volumes ...api.VolumeAssignment) {
+	r.mu.Lock()
+	var volumeObjects []*api.VolumeAssignment
+	defer r.mu.Unlock()
+	ctx := context.Background()
+	logctx := log.WithModule(ctx, "CSI Volume Add")
+	for _, volume := range volumes {
+		v := volume.Copy()
+		log.G(logctx).Debugf("Add Volume:%v", volume.VolumeID)
+
+		r.m[volume.VolumeID] = v
+		volumeObjects = append(volumeObjects, v)
+	}
+	go r.addVolumes(volumeObjects)
+}
+
+func (r *volumes) addVolumes(volumeObjects []*api.VolumeAssignment) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	ctx := context.Background()
+	logctx := log.WithModule(ctx, "CSI Volume Add")
+	for _, v := range volumeObjects {
+		retry := 0
+		for retry < MaxRetry {
+			err := r.plugin.NodeStageVolume(ctx, v)
+			if err == nil {
+				log.G(logctx).Debugf("Volume:%v staged successfully", v.VolumeID)
+				r.publishVolume(v)
+				break //break retry loop
+			}
+			log.G(logctx).WithError(err).Errorf("Volume stage failed for volID=%v, retrying...", v.VolumeID)
+			retry++
```

Comment:
Additionally, the CSI spec says that failures should be retried with "exponential backoff".